### PR TITLE
Prototypes manager

### PR DESCRIPTION
**Pull request by TomMalbran**
I moved all the prototypes from the world and moved them into its own classes. This will be useful when having multiple worlds, as they can all share the same prototypes.

Right now each prototype class is loading the files, but when we change how mods work, the file loading will be done by the mod manager, and it will just pass the file contents to the prototypes classes for parsing.

There are still more prototypes outside of the world (like the tile types) that should be moved. But for now I think that this is already a big change.

### CODE REVIEW

File: Assets/Scripts/Models/Prototypes/BasePrototypes.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public Dictionary<string, T>.KeyCollection Keys
+    {
+        get
+        {
+            return prototypes.Keys;
+        }
+    }
+
+    /// <summary>
+    /// Returns the prototype Keys.
```
Comment (Grenkin1988) [inline Assets/Scripts/Models/Prototypes/BasePrototypes.cs]: Values :-)

Comment (TomMalbran) [inline Assets/Scripts/Models/Prototypes/BasePrototypes.cs]: Woops. Not sure if that was me, or Mono that failed. hehe

**Comment by NogginBops** [comment]
This! 🙏

**Comment by Mizipzor** [comment]
This branch has conflicts that must be resolved, probably due to the huge StyleCop cleanup that just happened. Fix them and remove the `[Conflicts]` tag. :)

**Comment by TomMalbran** [comment]
Conflicts fixed. This time it was harder.

**Comment by Fatalexceptionz** [comment]
Seems to do what it says on the tin -- it runs, and my quick skim through the code didn't turn up anything glaringly obvious, though I'm getting sleep deprived, so...
Looks a lot cleaner and self-descriptive than it was.
BuildModeController could do with a cleanup next time someone touches it.
Could maybe be generalised even more to cut down on the boilerplate code, but, eh..

**Comment by TomMalbran** [comment]
@Fatalexceptionz I agree. I don't like how the BuildMode / Mose controllers combo works. They both depend on each other. I would like it to work in a way where the BuildModeController would ask the MouseController to call it back if something happens, so that the mouse controller wouldn't need to call it manually. And similar with a new SelectionModeController. But does seems like a big refactor that should be done on its own.

**Comment by vogonistic** [comment]
There are merge conflicts. Also remember to check if there are StyleCop warnings before removing `[conflict]` from the title.

**Comment by TomMalbran** [comment]
Conflicts fixed

**Comment by vogonistic** [comment]
Merged!

**Pull request merged**